README.md
# Kubernetes Certificate Signing Request 

[![CircleCI](https://circleci.com/gh/JulienBalestra/kube-csr.svg?style=svg)](https://circleci.com/gh/JulienBalestra/kube-csr) [![Build Status](https://travis-ci.org/JulienBalestra/kube-csr.svg?branch=master)](https://travis-ci.org/JulienBalestra/kube-csr) [![Go Report Card](https://goreportcard.com/badge/github.com/JulienBalestra/kube-csr)](https://goreportcard.com/report/github.com/JulienBalestra/kube-csr) [![Docker Repository on Quay](https://quay.io/repository/julienbalestra/kube-csr/status "Docker Repository on Quay")](https://quay.io/repository/julienbalestra/kube-csr)

## Table of Contents
- [Issue](#issue)
- [Garbage Collector](#garbage-collector---gc)
- [Demo](#demo)
- [Container image](#container-image)
- [Command line](#command-line)
    * [InCluster](#in-cluster)
- [Library](#library)
- [Features Enhancement](#features---enhancement)

## Issue

All in one:
* query Kubernetes service to get potential subject alternative name (SAN)
    * ClusterIP
    * ExternalIPs
    * LoadBalancerIP
    * ExternalName
* generate
    * Private Key - **stay on disk**
    * Certificate Signing Request (CSR)
* submit the generated CSR
* approve the submitted CSR
* fetch the generated certificate
* delete the kubernetes csr resource


But you can also choose to select the steps you want to execute.

For example, you can do the following actions:
* generate the PK, CSR
* submit the generated CSR
* <s>approve the submitted CSR</s>
* fetch the generated **externally approved** certificate
* <s>delete the kubernetes csr resource</s>

![diagram](docs/diagram.svg)


**Renew**

You can choose to add a sidecar to proceed to a periodic renew with, at least the following operations:
* submit the generated CSR
* fetch the generated **externally approved** certificate

A common scenario could be to automatically approve the submitted CSR like:
* use the previously generated files (private_key, csr)
* submit the generated CSR
* approve the submitted CSR
* fetch the generated certificate


## Garbage collector - gc

Delete any Kubernetes csr resources who meets the chosen requirements:

* After the grace period
    * Annotated as fetched
    * Denied

The garbage collector can be daemonized with the adapted flags.

When daemonised, it exposes a prometheus endpoint with the associated [metrics](./docs/metrics.csv) and a [pprof](https://golang.org/pkg/net/http/pprof/) endpoint.

## Demo

[![asciicast](https://asciinema.org/a/uIh0ujCiRiWJ6NOyLcEf369vq.png)](https://asciinema.org/a/uIh0ujCiRiWJ6NOyLcEf369vq)

## Container image

Available at `quay.io/julienbalestra/kube-csr:latest`

The tag `latest` is up to date with master.

Please, have a look to the [release  page](https://github.com/JulienBalestra/kube-csr/releases) to get a more stable image tag.

## Command line

Command line example:
```text
$ ./kube-csr issue etcd --generate --submit --approve --fetch --subject-alternative-names 192.168.1.1,example.com --kubeconfig-path ~/.kube/config

I0610 21:07:49.737337    5259 generate.go:59] Added IP address 192.168.1.1
I0610 21:07:49.737541    5259 generate.go:64] Added DNS name example.com
I0610 21:07:49.737547    5259 generate.go:73] CSR with 1 DNS names and 1 IP addresses
I0610 21:07:49.737553    5259 generate.go:94] Generating CSR with CN=etcd
I0610 21:07:49.740107    5259 write.go:57] Wrote RSA PRIVATE KEY to /home/jb/go/src/github.com/JulienBalestra/kube-csr/kube-csr.private_key
I0610 21:07:49.740142    5259 write.go:57] Wrote CERTIFICATE REQUEST to /home/jb/go/src/github.com/JulienBalestra/kube-csr/kube-csr.csr
I0610 21:07:49.750902    5259 submit.go:93] Successfully created csr/etcd-haf uid: 9163ed79-6ce1-11e8-a2ef-5404a66983a9
I0610 21:07:49.750932    5259 approve.go:38] Approving csr/etcd-haf ...
I0610 21:07:49.753173    5259 approve.go:49] csr/etcd-haf is approved
I0610 21:07:49.753192    5259 fetch.go:97] Start polling for certificate of csr/etcd-haf, every 1s, timeout after 10s
I0610 21:07:50.759603    5259 fetch.go:128] Certificate successfully fetched, writing 1200 chars to /home/jb/go/src/github.com/JulienBalestra/kube-csr/kube-csr.certificate
```

The `--override` flag allows to delete and re-submit an already submitted CSR.

To get the following files:
```text
kube-csr.certificate kube-csr.csr kube-csr.private_key
```

```text
$ openssl x509 -in kube-csr.certificate -text -noout

Certificate:
        Issuer: CN = p8s
        Subject: CN = etcd
            X509v3 Subject Alternative Name: 
                DNS:etcd-0.default.svc.cluster.local, IP Address:192.168.1.1
```

Observe in the controller-manager logs:
```text
$ kubectl logs po/kube-controller-manager -n kube-system

[INFO] signed certificate with serial number [...]
```

Have a look the the command line documentation [here](docs/kube-csr.md)

### In cluster

The current Kubernetes setup is deployed with [pupernetes](https://github.com/DataDog/pupernetes)
In this example, the all in one [etcd example](examples/etcd.yaml) is used.

Current cluster:
```text
$ kubectl get svc,deploy,ds,po --all-namespaces

NAMESPACE     NAME         TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)         AGE
default       kubernetes   ClusterIP   192.168.254.1   <none>        443/TCP         1m
kube-system   coredns      ClusterIP   192.168.254.2   <none>        53/UDP,53/TCP   1m

NAMESPACE     NAME      DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
kube-system   coredns   1         1         1            1           1m

NAMESPACE     NAME             DESIRED   CURRENT   READY     UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
kube-system   kube-proxy       1         1         1         1            1           <none>          1m
kube-system   kube-scheduler   1         1         1         1            1           <none>          1m

NAMESPACE     NAME                       READY     STATUS    RESTARTS   AGE
kube-system   coredns-747dbcf5df-zllhm   1/1       Running   0          1m
kube-system   kube-controller-manager    1/1       Running   0          1m
kube-system   kube-proxy-2z9vw           1/1       Running   0          1m
kube-system   kube-scheduler-v8lwc       1/1       Running   0          1m
``` 

Apply the manifests:
```text
$ kubectl apply -f examples/etcd.yaml 

serviceaccount "etcd" created
clusterrole.rbac.authorization.k8s.io "system:etcd" created
clusterrolebinding.rbac.authorization.k8s.io "system:etcd" created
statefulset.apps "etcd" created
job.batch "etcdctl" created
service "etcd" created
```

Produce:
```text
$ kubectl get csr,po --show-all

NAME                                              AGE       REQUESTOR                            CONDITION
csr/etcd-0-140d1b66-6ce3-11e8-a2ef-5404a66983a9   1m        system:serviceaccount:default:etcd   Approved,Issued
csr/etcd-1-173e7ddd-6ce3-11e8-a2ef-5404a66983a9   1m        system:serviceaccount:default:etcd   Approved,Issued
csr/etcd-2-1a6a22a1-6ce3-11e8-a2ef-5404a66983a9   59s       system:serviceaccount:default:etcd   Approved,Issued

NAME        READY     STATUS    RESTARTS   AGE
po/etcd-0   1/1       Running   0          1m
po/etcd-1   1/1       Running   0          1m
po/etcd-2   1/1       Running   0          1m
```

Observe the logs of the init container kube-csr:
```text
$ kubectl logs etcd-0 kube-csr

I0610 19:18:39.760447       1 generate.go:59] Added IP address 172.17.0.2
I0610 19:18:39.760763       1 generate.go:64] Added DNS name etcd-0.etcd.default.svc.cluster.local
I0610 19:18:39.760768       1 generate.go:64] Added DNS name etcd.default.svc.cluster.local

releasenotes.md
- [v0.5.0](#v0.5.0)
- [v0.4.0](#v0.4.0)
- [v0.3.0](#v0.3.0)
- [v0.2.0](#v0.2.0)

## v0.5.0

### Features
* Introduce the renew feature #54
* Move the main to the root #48
* Add logging messages in the operation package #45

### Bugfix
* Dedup and sort SAN, validate all DNS names #51

### Other
* ci: bump p8s #55
* Upgrade dep to 0.5 #52
* Metrics-server is self registered #50
* ci: bump p8s to 0.7.0 #46
* Add the metrics-server as example #44
* example: add the query, test the issue.go in e2e #43

## v0.4.0

### Features
* Add an expired purge capability #40
* Include Kubernetes services DNS and IPs in SAN #41

### Other
* Add unittests in generate and pemio #39
* Improve ci #38
* Upgrade to p8s 0.6.1 #35

## v0.3.0

### Features
* Add a pprof http handler #31
* Add a prometheus exporter handler #23

### Bugfix
* tickers: must be non zero #22

### Other
* Create a ToC #30
* Add hyperkube version 1.11.0 #29
* Add hyperkube versions #26
* Add travis status badge #25
* Introduce pupernetes as e2e testing engine. #24
* readme: update the diagram, replace purge by delete #21
* Update the diagram and add the gc to the docs #20

## v0.2.0

### Features
* Introduce the Garbage Collect command #19
* Create a delete flag #15
* Be able to skip the fetch annotation updates #13
* Set annotations to track the activity of fetch #12

### Bugfix
* Build from the current working directory #18
* Fetch fail on denied #11

### Other
* Add a probe to the etcd example #17
* Add a misspel tool #16
* Create a diagram on the README #14


cmd/cmd.go
package cmd

import (
	"flag"
	"fmt"
	"os"
	"path"
	"strconv"
	"time"

	"github.com/golang/glog"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"

	"github.com/JulienBalestra/kube-csr/pkg/operation"
	"github.com/JulienBalestra/kube-csr/pkg/operation/approve"
	"github.com/JulienBalestra/kube-csr/pkg/operation/fetch"
	"github.com/JulienBalestra/kube-csr/pkg/operation/generate"
	"github.com/JulienBalestra/kube-csr/pkg/operation/purge"
	"github.com/JulienBalestra/kube-csr/pkg/operation/query"
	"github.com/JulienBalestra/kube-csr/pkg/operation/submit"
	"github.com/JulienBalestra/kube-csr/pkg/renew"
)

const (
	programName = "kube-csr"
)

var (
	viperConfig = viper.New()
)

// NewCommand creates a cobra command to be consumed in the main package
func NewCommand() (*cobra.Command, *int) {
	var verbose int
	var exitCode int

	rootCommand := &cobra.Command{
		Use:   programName,
		Short: "Use this command to manage Kubernetes certificates",
		PersistentPreRun: func(cmd *cobra.Command, args []string) {
			flag.Lookup("alsologtostderr").Value.Set("true")
			flag.Lookup("v").Value.Set(strconv.Itoa(verbose))
		},
	}
	rootCommand.PersistentFlags().IntVarP(&verbose, "verbose", "v", 0, "verbose level")
	viperConfig.SetDefault("kubeconfig-path", "")
	rootCommand.PersistentFlags().String("kubeconfig-path", viperConfig.GetString("kubeconfig-path"), "Kubernetes config path, leave empty for inCluster config")
	viperConfig.BindPFlag("kubeconfig-path", rootCommand.PersistentFlags().Lookup("kubeconfig-path"))

	garbageCommandName := fmt.Sprintf("%s gc", programName)
	garbageCommand := &cobra.Command{
		Use:        "garbage-collect",
		Args:       cobra.ExactArgs(0),
		Aliases:    []string{"gc"},
		SuggestFor: []string{"purge", "delete", "remove", "del", "rm"},
		Short:      "Garbage collect Kubernetes certificates on different parameters",
		Example: fmt.Sprintf(`
# Garbage collect all csr already fetched with a grace period of 12 hours
%s --fetched --grace-period=12h

# Garbage collect all csr denied with a grace period of 15 minutes
%s --fetched --grace-period=15m

# Garbage collect now all csr already fetched
%s --fetched --grace-period=0s

# Garbage collect every 10min all csr already fetched with a grace period of 1 hour
%s --fetched --daemon polling-period=10m --grace-period=1h
`,
			garbageCommandName,
			garbageCommandName,
			garbageCommandName,
			garbageCommandName,
		),
		Run: func(cmd *cobra.Command, args []string) {
			if !viperConfig.GetBool("denied") &&
				!viperConfig.GetBool("fetched") &&
				!viperConfig.GetBool("expired") {
				glog.Errorf("Must choose at least one flag: --denied, --fetched, --expired")
				exitCode = 1
				return
			}
			gc, err := newGarbageCollector()
			if err != nil {
				exitCode = 1
				return
			}
			if viperConfig.GetBool("daemon") {
				err = gc.GarbageCollectLoop()
			} else {
				err = gc.GarbageCollect()
			}
			if err != nil {
				exitCode = 2
				return
			}
		},
	}
	rootCommand.AddCommand(garbageCommand)

	// gc safety period
	viperConfig.SetDefault("grace-period", time.Hour*48)
	garbageCommand.PersistentFlags().Duration("grace-period", viperConfig.GetDuration("grace-period"), "duration to wait before deleting Kubernetes csr objects")
	viperConfig.BindPFlag("grace-period", garbageCommand.PersistentFlags().Lookup("grace-period"))

	// gc functions
	viperConfig.SetDefault("denied", false)
	garbageCommand.PersistentFlags().Bool("denied", viperConfig.GetBool("denied"), "delete any denied Kubernetes csr")
	viperConfig.BindPFlag("denied", garbageCommand.PersistentFlags().Lookup("denied"))

	viperConfig.SetDefault("fetched", false)
	garbageCommand.PersistentFlags().Bool("fetched", viperConfig.GetBool("fetched"), fmt.Sprintf("delete any already fetched Kubernetes csr, the state is tracked with kube-annotations %q", fetch.KubeCSRFetchedAnnotationPrefix))
	viperConfig.BindPFlag("fetched", garbageCommand.PersistentFlags().Lookup("fetched"))

	viperConfig.SetDefault("expired", false)
	garbageCommand.PersistentFlags().Bool("expired", viperConfig.GetBool("expired"), fmt.Sprintf("delete any Kubernetes csr with an expired certificate"))
	viperConfig.BindPFlag("expired", garbageCommand.PersistentFlags().Lookup("expired"))

	// daemon flags
	pollingPeriod, daemon := "polling-period", "daemon"
	viperConfig.SetDefault(pollingPeriod, time.Minute*10)
	garbageCommand.PersistentFlags().Duration(pollingPeriod, viperConfig.GetDuration(pollingPeriod), fmt.Sprintf("duration to wait between each gc call, paired with --%s", daemon))
	viperConfig.BindPFlag(pollingPeriod, garbageCommand.PersistentFlags().Lookup(pollingPeriod))

	viperConfig.SetDefault(daemon, false)
	garbageCommand.PersistentFlags().Bool(daemon, viperConfig.GetBool(daemon), fmt.Sprintf("continually gc Kubernetes csr, paired with --%s", pollingPeriod))
	viperConfig.BindPFlag(daemon, garbageCommand.PersistentFlags().Lookup(daemon))

	viperConfig.SetDefault("disable-prometheus-exporter", false)
	garbageCommand.PersistentFlags().Bool("disable-prometheus-exporter", viperConfig.GetBool("disable-prometheus-exporter"), fmt.Sprintf("disable /metrics, paired with --%s", daemon))
	viperConfig.BindPFlag("disable-prometheus-exporter", garbageCommand.PersistentFlags().Lookup("disable-prometheus-exporter"))

	viperConfig.SetDefault("prometheus-exporter-bind", "0.0.0.0:8484")
	garbageCommand.PersistentFlags().Bool("prometheus-exporter-bind", viperConfig.GetBool("prometheus-exporter-bind"), fmt.Sprintf("prometheus exporter bind address, paired with --%s", daemon))
	viperConfig.BindPFlag("prometheus-exporter-bind", garbageCommand.PersistentFlags().Lookup("prometheus-exporter-bind"))

	// issue command
	issueCommandName := fmt.Sprintf("%s issue", programName)
	issueCommand := &cobra.Command{
		Use:        "issue",
		Aliases:    []string{"i"},
		SuggestFor: []string{"issuer", "isue", "issu"},
		Short:      "Use this command to generate, approve, fetch and self-delete Kubernetes certificates",
		Args:       cobra.ExactArgs(1),
		Example: fmt.Sprintf(`
# generate the private key and the csr
%s my-app --generate

# generate the private key, the csr and then submit the csr
%s my-app --generate --submit

# generate the private key, the csr, submit and approve the csr
%s my-app --generate --submit --approve

# Generate the private key, the csr, submit, approve and fetch the csr
%s my-app --generate --submit --approve --fetch
%s my-app -gsaf
%s my-app -gsaf --subject-alternative-names 192.168.1.1,etcd-0.default.svc.cluster.local

# Generate the private key, the csr, submit and fetch the csr when externally approved
%s my-app --generate --submit --fetch --fetch-interval 10s --fetch-timeout 10m

# Generate the private key, the csr, submit, approve, fetch and delete the csr
%s my-app --generate --submit --approve --fetch --delete 

# Generate the private key, the csr, submit, approve and fetch the csr. Override any existing and use a kubeconfig
%s my-app -gsaf --override --kubeconfig-path ~/.kube/config

# Execute all steps with a custom kubernetes csr name
%s skydns --csr-name kv-etcd -gsafd --override --kubeconfig-path ~/.kube/config
`,
			// TODO use go template here
			issueCommandName,
			issueCommandName,
			issueCommandName,
			issueCommandName,
			issueCommandName,
			issueCommandName,
			issueCommandName,
			issueCommandName,
			issueCommandName,
			issueCommandName,
		),
		Run: func(cmd *cobra.Command, args []string) {
			if !viperConfig.GetBool("generate") &&
				!viperConfig.GetBool("renew") &&
				!viperConfig.GetBool("submit") &&
				!viperConfig.GetBool("approve") &&
				!viperConfig.GetBool("fetch") &&
				!viperConfig.GetBool("delete") {
				glog.Errorf("Must choose at least one flag: --generate, --submit, --approve, --fetch, --delete")
				exitCode = 1
				return
			}
			// build common name and csr name
			commonName := args[0]
			csrName, err := generateCertificateSigningRequestName(commonName)
			if err != nil {
				exitCode = 1
				return
			}

			csrConfig, err := newCSRConfig(commonName, csrName)
			if err != nil {
				exitCode = 1
				return
			}
			var querier *query.Query
			var generator *generate.Generator
			var submitter *submit.Submit
			var approval *approve.Approval
			var fetcher *fetch.Fetch
			var purger *purge.Purge

			svcToQuery := viperConfig.GetStringSlice("query-svc")
			if len(svcToQuery) > 0 {
				querier, err = newQuery(svcToQuery)
				if err != nil {
					exitCode = 1
					return
				}
			}

			if viperConfig.GetBool("generate") {
				generator = generate.NewGenerator(csrConfig)
			}
			if viperConfig.GetBool("submit") || viperConfig.GetBool("renew") {
				submitter, err = newSubmitClient()
				if err != nil {
					exitCode = 1
					return
				}
			}
			if viperConfig.GetBool("approve") {
				approval, err = newApproveClient()
				if err != nil {
					exitCode = 1
					return
				}
			}
			if viperConfig.GetBool("fetch") || viperConfig.GetBool("renew") {
				fetcher, err = newFetchClient()
				if err != nil {
					exitCode = 1
					return
				}
			}
			if viperConfig.GetBool("delete") {
				purger, err = newDeleteClient()
				if err != nil {
					exitCode = 1
					return
				}
			}
			op := operation.NewOperation(
				&operation.Config{
					SourceConfig: csrConfig,
					Query:        querier,
					Generate:     generator,
					Submit:       submitter,
					Approve:      approval,
					Fetch:        fetcher,
					Purge:        purger,
				},
			)
			if !viperConfig.GetBool("renew") {
				err = op.Run()
				if err != nil {
					glog.Errorf("Unexpected error: %v", err)
					exitCode = 2
					return
				}
				return
			}
			re, err := newRenew(op)
			if err != nil {
				exitCode = 1
				return
			}
			err = re.Renew()
			if err != nil {
				glog.Errorf("Unexpected error: %v", err)
				exitCode = 2
				return
			}
		},
	}
	rootCommand.AddCommand(issueCommand)

	// all
	viperConfig.SetDefault("csr-name", "")
	issueCommand.PersistentFlags().String("csr-name", viperConfig.GetString("csr-name"), "Kubernetes CSR name, leave empty for CN-hostname")
	viperConfig.BindPFlag("csr-name", issueCommand.PersistentFlags().Lookup("csr-name"))

	viperConfig.SetDefault("hostname", "")
	issueCommand.PersistentFlags().String("hostname", viperConfig.GetString("hostname"), "Hostname, leave empty to fulfill with hostname")
	viperConfig.BindPFlag("hostname", issueCommand.PersistentFlags().Lookup("hostname"))

	viperConfig.SetDefault("override", false)
	issueCommand.PersistentFlags().Bool("override", viperConfig.GetBool("override"), "Override any existing file pem and k8s csr resource")
	viperConfig.BindPFlag("override", issueCommand.PersistentFlags().Lookup("override"))

	// query
	viperConfig.SetDefault("query-svc", nil)
	issueCommand.PersistentFlags().StringSliceP("query-svc", "q", viperConfig.GetStringSlice("query-svc"), "Query the kube-apiserver services to get additional SAN (namespaceName/serviceName) comma separated")
	viperConfig.BindPFlag("query-svc", issueCommand.PersistentFlags().Lookup("query-svc"))

	viperConfig.SetDefault("query-interval", time.Second*2)
	issueCommand.PersistentFlags().Duration("query-interval", viperConfig.GetDuration("query-interval"), "Polling interval for kube-service query")
	viperConfig.BindPFlag("query-interval", issueCommand.PersistentFlags().Lookup("query-interval"))

	viperConfig.SetDefault("query-timeout", time.Second*20)
	issueCommand.PersistentFlags().Duration("query-timeout", viperConfig.GetDuration("query-timeout"), "Polling timeout for kube-service query")
	viperConfig.BindPFlag("query-timeout", issueCommand.PersistentFlags().Lookup("query-timeout"))

	// generate
	viperConfig.SetDefault("generate", false)
	issueCommand.PersistentFlags().BoolP("generate", "g", viperConfig.GetBool("generate"), "Generate CSR")
	viperConfig.BindPFlag("generate", issueCommand.PersistentFlags().Lookup("generate"))

	viperConfig.SetDefault("rsa-bits", 2048)
	issueCommand.PersistentFlags().String("rsa-bits", viperConfig.GetString("rsa-bits"), "RSA bits for the private key")
	viperConfig.BindPFlag("rsa-bits", issueCommand.PersistentFlags().Lookup("rsa-bits"))

	viperConfig.SetDefault("subject-alternative-names", nil)
	issueCommand.PersistentFlags().StringSlice("subject-alternative-names", viperConfig.GetStringSlice("subject-alternative-names"), "Subject Alternative Names (SANs) comma separated")
	viperConfig.BindPFlag("subject-alternative-names", issueCommand.PersistentFlags().Lookup("subject-alternative-names"))

	// generate - private key
	viperConfig.SetDefault("private-key-perm", 0600)

	viperConfig.SetDefault("private-key-file", "kube-csr.private_key")
	issueCommand.PersistentFlags().String("private-key-file", viperConfig.GetString("private-key-file"), "Private key file target")
	viperConfig.BindPFlag("private-key-file", issueCommand.PersistentFlags().Lookup("private-key-file"))

	viperConfig.SetDefault("load-private-key", false)
	issueCommand.PersistentFlags().Bool("load-private-key", viperConfig.GetBool("load-private-key"), "Load the private key file instead of generating one")
	viperConfig.BindPFlag("load-private-key", issueCommand.PersistentFlags().Lookup("load-private-key"))

	// generate - csr
	viperConfig.SetDefault("csr-perm", 0600)

	viperConfig.SetDefault("csr-file", "kube-csr.csr")
	issueCommand.PersistentFlags().String("csr-file", viperConfig.GetString("csr-file"), "Certificate Signing Request file target")
	viperConfig.BindPFlag("csr-file", issueCommand.PersistentFlags().Lookup("csr-file"))

	// submit
	viperConfig.SetDefault("submit", false)
	issueCommand.PersistentFlags().BoolP("submit", "s", viperConfig.GetBool("submit"), "Submit the CSR")
	viperConfig.BindPFlag("submit", issueCommand.PersistentFlags().Lookup("submit"))

	// approve
	viperConfig.SetDefault("approve", false)
	issueCommand.PersistentFlags().BoolP("approve", "a", viperConfig.GetBool("approve"), "Approve the CSR")
	viperConfig.BindPFlag("approve", issueCommand.PersistentFlags().Lookup("approve"))

	// fetch
	viperConfig.SetDefault("fetch", false)
	viperConfig.SetDefault("certificate-perm", 0600)

	issueCommand.PersistentFlags().BoolP("fetch", "f", viperConfig.GetBool("fetch"), "Fetch the CSR")
	viperConfig.BindPFlag("fetch", issueCommand.PersistentFlags().Lookup("fetch"))

	viperConfig.SetDefault("certificate-file", "kube-csr.certificate")
	issueCommand.PersistentFlags().String("certificate-file", viperConfig.GetString("certificate-file"), "Certificate file target")
	viperConfig.BindPFlag("certificate-file", issueCommand.PersistentFlags().Lookup("certificate-file"))

	viperConfig.SetDefault("fetch-interval", time.Second*1)
	issueCommand.PersistentFlags().Duration("fetch-interval", viperConfig.GetDuration("fetch-interval"), "Polling interval for certificate fetching")
	viperConfig.BindPFlag("fetch-interval", issueCommand.PersistentFlags().Lookup("fetch-interval"))

	viperConfig.SetDefault("fetch-timeout", time.Second*10)
	issueCommand.PersistentFlags().Duration("fetch-timeout", viperConfig.GetDuration("fetch-timeout"), "Polling timeout for certificate fetching")
	viperConfig.BindPFlag("fetch-timeout", issueCommand.PersistentFlags().Lookup("fetch-timeout"))

	viperConfig.SetDefault("skip-fetch-annotate", false)
	issueCommand.PersistentFlags().Bool("skip-fetch-annotate", viperConfig.GetBool("skip-fetch-annotate"), "Skip the update of annotations when successfully fetched the certificate")
	viperConfig.BindPFlag("skip-fetch-annotate", issueCommand.PersistentFlags().Lookup("skip-fetch-annotate"))

	// delete
	viperConfig.SetDefault("delete", false)
	issueCommand.PersistentFlags().BoolP("delete", "d", viperConfig.GetBool("delete"), "Delete the given CSR from the kube-apiserver")
	viperConfig.BindPFlag("delete", issueCommand.PersistentFlags().Lookup("delete"))

	// renew
	viperConfig.SetDefault("renew", false)
	issueCommand.PersistentFlags().Bool("renew", viperConfig.GetBool("renew"), "Renew")
	viperConfig.BindPFlag("renew", issueCommand.PersistentFlags().Lookup("renew"))

	viperConfig.SetDefault("renew-exit", false)
	issueCommand.PersistentFlags().Bool("renew-exit", viperConfig.GetBool("renew-exit"), "Exit 0 after a successful renew")
	viperConfig.BindPFlag("renew-exit", issueCommand.PersistentFlags().Lookup("renew-exit"))

	viperConfig.SetDefault("renew-command", "")
	issueCommand.PersistentFlags().String("renew-command", viperConfig.GetString("renew-command"), "Command to execute after a successful renew (using /bin/sh as interpreter)")
	viperConfig.BindPFlag("renew-command", issueCommand.PersistentFlags().Lookup("renew-command"))

	viperConfig.SetDefault("renew-threshold", time.Hour)
	issueCommand.PersistentFlags().Duration("renew-threshold", viperConfig.GetDuration("renew-threshold"), "Renew expiration threshold")
	viperConfig.BindPFlag("renew-threshold", issueCommand.PersistentFlags().Lookup("renew-threshold"))

	viperConfig.SetDefault("renew-check-interval", time.Minute*15)
	issueCommand.PersistentFlags().Duration("renew-check-interval", viperConfig.GetDuration("renew-check-interval"), "Interval between check of the certificate expiration")
	viperConfig.BindPFlag("renew-check-interval", issueCommand.PersistentFlags().Lookup("renew-check-interval"))

	issueCommand.PersistentFlags().Bool("disable-prometheus-exporter", viperConfig.GetBool("disable-prometheus-exporter"), "disable /metrics, paired with --renew")
	viperConfig.BindPFlag("disable-prometheus-exporter", garbageCommand.PersistentFlags().Lookup("disable-prometheus-exporter"))

	issueCommand.PersistentFlags().Bool("prometheus-exporter-bind", viperConfig.GetBool("prometheus-exporter-bind"), "prometheus exporter bind address, paired with --renew")
	viperConfig.BindPFlag("prometheus-exporter-bind", garbageCommand.PersistentFlags().Lookup("prometheus-exporter-bind"))
	return rootCommand, &exitCode
}

func generateCertificateSigningRequestName(commonName string) (string, error) {
	csrName := viperConfig.GetString("csr-name")
	if csrName != "" {
		return csrName, nil
	}

	hostname := viperConfig.GetString("hostname")
	if hostname != "" {
		return fmt.Sprintf("%s-%s", commonName, hostname), nil
	}

	hostname, err := os.Hostname()
	if err != nil {
		glog.Errorf("Cannot get hostname: %v", err)
		return "", err
	}
	return fmt.Sprintf("%s-%s", commonName, hostname), nil
}

func newCSRConfig(commonName, csrName string) (*generate.Config, error) {
	wd, err := os.Getwd()
	if err != nil {
		glog.Errorf("Unexpected error: %v", err)
		return nil, err
	}

	privateKeyPath := viperConfig.GetString("private-key-file")
	if !path.IsAbs(privateKeyPath) {
		privateKeyPath = path.Join(wd, privateKeyPath)
	}

	csrPath := viperConfig.GetString("csr-file")
	if !path.IsAbs(csrPath) {
		csrPath = path.Join(wd, csrPath)
	}
	return &generate.Config{
		Name:       csrName,
		Override:   viperConfig.GetBool("override"),
		CommonName: commonName,
		Hosts:      viperConfig.GetStringSlice("subject-alternative-names"),
		RSABits:    viperConfig.GetInt("rsa-bits"),

		LoadPrivateKey:       viperConfig.GetBool("load-private-key"),
		PrivateKeyABSPath:    privateKeyPath,
		PrivateKeyPermission: os.FileMode(viperConfig.GetInt("private-key-perm")),

		CSRABSPath:    csrPath,
		CSRPermission: os.FileMode(viperConfig.GetInt("csr-perm")),
	}, nil
}

func newSubmitClient() (*submit.Submit, error) {
	s, err := submit.NewSubmitter(
		viperConfig.GetString("kubeconfig-path"),
		&submit.Config{
			Override: viperConfig.GetBool("override"),
		},
	)
	if err != nil {
		return nil, err
	}
	return s, nil
}

func newApproveClient() (*approve.Approval, error) {
	s, err := approve.NewApproval(viperConfig.GetString("kubeconfig-path"))
	if err != nil {
		return nil, err
	}
	return s, nil
}

func newFetchClient() (*fetch.Fetch, error) {
	wd, err := os.Getwd()
	if err != nil {
		glog.Errorf("Unexpected error: %v", err)
		return nil, err
	}

	crtPath := viperConfig.GetString("certificate-file")
	if !path.IsAbs(crtPath) {
		crtPath = path.Join(wd, crtPath)
	}

	annotate := !viperConfig.GetBool("skip-fetch-annotate")
	if annotate && viperConfig.GetBool("delete") {
		// useless to annotate a kube resource just deleted after
		glog.V(0).Infof("As configured with delete, ignoring the annotation operations during the fetch")
		annotate = false
	}
	conf := &fetch.Config{
		Override:              viperConfig.GetBool("override"),
		PollingInterval:       viperConfig.GetDuration("fetch-interval"),
		PollingTimeout:        viperConfig.GetDuration("fetch-timeout"),
		CertificatePermission: os.FileMode(viperConfig.GetInt("certificate-perm")),
		CertificateABSPath:    crtPath,
		Annotate:              annotate,
	}
	f, err := fetch.NewFetcher(viperConfig.GetString("kubeconfig-path"), conf)
	if err != nil {
		return nil, err
	}
	return f, nil
}

func newDeleteClient() (*purge.Purge, error) {
	s, err := purge.NewPurge(viperConfig.GetString("kubeconfig-path"), nil)
	if err != nil {
		return nil, err
	}
	return s, nil
}

func newGarbageCollector() (*purge.Purge, error) {
	conf := purge.NewPurgeConfig(viperConfig.GetDuration("grace-period"))
	if viperConfig.GetBool("denied") {
		conf.ShouldGC = append(conf.ShouldGC, purge.IsConditionDenied)
	}
	if viperConfig.GetBool("fetched") {
		conf.ShouldGC = append(conf.ShouldGC, purge.IsAnnotationFetched)
	}
	if viperConfig.GetBool("expired") {
		conf.ShouldGC = append(conf.ShouldGC, purge.IsCertificateExpired)
	}
	conf.PollingPeriod = viperConfig.GetDuration("polling-period")
	if !viperConfig.GetBool("disable-prometheus-exporter") {
		conf.PrometheusExporterBindAddress = viperConfig.GetString("prometheus-exporter-bind")
	}
	p, err := purge.NewPurge(viperConfig.GetString("kubeconfig-path"), conf)
	if err != nil {
		return nil, err
	}
	return p, nil
}

func newQuery(svcToQuery []string) (*query.Query, error) {
	q, err := query.NewQuery(viperConfig.GetString("kubeconfig-path"), svcToQuery, &query.Config{
		PollingTimeout:  viperConfig.GetDuration("query-timeout"),
		PollingInterval: viperConfig.GetDuration("query-interval"),
	})
	if err != nil {
		return nil, err
	}
	return q, nil
}

func newRenew(operation *operation.Operation) (*renew.Renew, error) {
	conf := &renew.Config{
		Operation:                operation,
		RenewThreshold:           viperConfig.GetDuration("renew-threshold"),
		ExitOnRenew:              viperConfig.GetBool("renew-exit"),
		GenerateNewKubernetesCSR: !viperConfig.GetBool("override"),
		RenewCommand:             viperConfig.GetString("renew-command"),
		RenewCheckInterval:       viperConfig.GetDuration("renew-check-interval"),
	}
	if !viperConfig.GetBool("disable-prometheus-exporter") {
		conf.PrometheusExporterBindAddress = viperConfig.GetString("prometheus-exporter-bind")
	}
	return renew.NewRenewer(viperConfig.GetString("kubeconfig-path"), conf)
}
